### PR TITLE
[Android] Fix TextInputType.none for devices with physical keyboard

### PR DESCRIPTION
**Pull request by bartcone**
## Description

This PR fixes an issue where keystrokes aren't received on Android devices with physical keyboards (e.g. rugged Zebra devices) when `keyboardType` is set to `TextInputType.none` on a `TextField`.

The logic in `setTextInputClient` and `canShowTextInput` created an `inputTarget` with `InputTarget.Type.NO_TARGET` which caused the [input connection to short circuit](https://github.com/flutter/engine/blob/main/shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java#L296) and not be established.

Bug introduction PR: https://github.com/flutter/engine/pull/26585

## Related Issue

https://github.com/flutter/flutter/issues/89983

## Unit Test Notes

- The existing `showTextInput_textInputTypeNone()` stays green after update.
- `inputConnection_textInputTypeNone()` updated to `assertNotNull`. I would make this more specific, but this is my first venture into the Flutter engine and don't know enough about those connection attributes.

## Demo

Video below with Zebra MC9300 device. This issue can also be reproduced in a standard android emulator. Simply add a `TextField`, configure `keyboardType` to be `TextInputType.none` and attempt to enter text after running and giving focus to textfield.

Before

https://github.com/flutter/engine/assets/1988098/348ca061-b8b9-4483-956e-0732c1238207

After

https://github.com/flutter/engine/assets/1988098/b65c7251-59b4-4c73-9b85-7ac03f47a7e4

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: shell/platform/android/test/io/flutter/plugin/editing/TextInputPluginTest.java [hunk 1]
```diff
@@ -1310,7 +1311,7 @@ public void inputConnection_textInputTypeNone() {
     InputConnection connection =
         textInputPlugin.createInputConnection(
             testView, mock(KeyboardManager.class), new EditorInfo());
-    assertEquals(connection, null);
+    assertNotNull(connection);
```
Comment (reidbaker) [inline shell/platform/android/test/io/flutter/plugin/editing/TextInputPluginTest.java:1314]: I don't think this pr has the correct level of testing for this bug.  I would expect either a new test that simulates some version of input and verifies it gets sent when the type is NONE or an existing test being modified from ensuring text was not sent to text being sent.

Comment (bartcone) [reply shell/platform/android/test/io/flutter/plugin/editing/TextInputPluginTest.java:1314]: Agreed. I'll try to find some time over the next few days to find a test case I can piggy back off of.

File: shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java [hunk 2]
```diff
@@ -387,16 +387,9 @@ public void sendTextInputAppPrivateCommand(@NonNull String action, @NonNull Bund
...
-    if (configuration == null || configuration.inputType == null) {
-      return true;
-    }
-    return configuration.inputType.type != TextInputChannel.TextInputType.NONE;
-  }
-
   @VisibleForTesting
   void showTextInput(View view) {
-    if (canShowTextInput()) {
+    if (configuration.inputType.type != TextInputChannel.TextInputType.NONE) {
```
Comment (johnmccutchan) [inline shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java:392]: the old code is checking for configuration being null and inputType being null. Why don't we need those checks anymore?

File: shell/platform/android/test/io/flutter/plugin/editing/TextInputPluginTest.java (hunk 1 above)
Comment (johnmccutchan) [reply shell/platform/android/test/io/flutter/plugin/editing/TextInputPluginTest.java:1314]: +1 to needing a proper test case that fails without your change to TextInputPlugin.java applied.

Comment (bartcone) [reply shell/platform/android/test/io/flutter/plugin/editing/TextInputPluginTest.java:1372]: @johnmccutchan @reidbaker added additional test case. I admittedly just copied the existing `inputConnection_createsActionFromEnter` test case and changed the `TextInputType` to `NONE` which caused a NullPointerException within the test case itself (source of the bug, no connection created). Obviously after updates, test case passes.

File: shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java (hunk 2 above)
Comment (bartcone) [reply shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java]: Added back.

**Review by reidbaker — APPROVED** [review]

**Comment by bartcone** [comment]
@LongCatIsLooong @reidbaker is the `[Linux linux_unopt]` check historically flakey? I've seen it fail twice now as this PR has evolved, but "re-running" has turned it green. Just wanted to call it out in case we needed to look into something further.

**Comment by reidbaker** [comment]
> @LongCatIsLooong @reidbaker is the `[Linux linux_unopt]` check historically flakey? I've seen it fail twice now as this PR has evolved, but "re-running" has turned it green. Just wanted to call it out in case we needed to look into something further.
> 
> e.g. https://github.com/flutter/engine/runs/21045950042

No it is not normally flakey for me but I think the issue might be https://github.com/flutter/flutter/issues/138197

File: shell/platform/android/test/io/flutter/plugin/editing/TextInputPluginTest.java (hunk 1 above)
Comment (johnmccutchan) [reply shell/platform/android/test/io/flutter/plugin/editing/TextInputPluginTest.java:1372]: Could you refactor the original test into a function that accepts TextInputType as an argument and avoid the code duplication?

**Review by reidbaker — APPROVED** [review]

**Comment by reidbaker** [comment]
Much cleaner after the refactor

File: shell/platform/android/test/io/flutter/plugin/editing/TextInputPluginTest.java (hunk 1 above)
Comment (johnmccutchan) [reply shell/platform/android/test/io/flutter/plugin/editing/TextInputPluginTest.java:1324]: Thanks for making the change!

**Review by johnmccutchan — APPROVED** [review]

**Comment by chinmaygarde** [comment]
The presub is failing on a formatting error.

**Comment by kishankumawat03** [comment]
I have upgraded my Flutter version to 3.19.6. but I am still having the issue. I am unable to receive data from the physical keyboard when keyboardType is set to TextInputType.none on a TextField.

**Comment by bartcone** [comment]
> I have upgraded my Flutter version to 3.19.6. but I am still having the issue. I am unable to receive data from the physical keyboard when keyboardType is set to TextInputType.none on a TextField.

https://github.com/flutter/flutter/issues/89983#issuecomment-1940424204

**Pull request merged**